pkg/imgpkg/plainimage/plain_image.go

```diff
@@ -9,27 +9,34 @@ import (
 	"github.com/cppforlife/go-cli-ui/ui"
 	regname "github.com/google/go-containerregistry/pkg/name"
 	regv1 "github.com/google/go-containerregistry/pkg/v1"
+	regremote "github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/types"
 	ctlimg "github.com/k14s/imgpkg/pkg/imgpkg/image"
 )
 
+type ImagesDescriptor interface {
+	Get(regname.Reference) (*regremote.Descriptor, error)
+}
+
 type PlainImage struct {
-	ref      string
-	registry ctlimg.ImagesMetadata
+	imagesDescriptor ImagesDescriptor
 
 	parsedRef    regname.Reference
 	parsedDigest string
 
 	fetchedImage regv1.Image
-	fetchedIndex regv1.ImageIndex
 }
 
-func NewPlainImage(ref string, imagesMetadata ctlimg.ImagesMetadata) *PlainImage {
-	return &PlainImage{ref: ref, registry: imagesMetadata}
-}
+func NewPlainImage(ref string, imgDescriptor ImagesDescriptor) *PlainImage {
+	parsedRef, err := regname.ParseReference(ref, regname.WeakValidation)
+	if err != nil {
+		panic(fmt.Sprintf("Expected valid Tag Ref: %s", err))
```

Comment:
true. true...

Ok, i went a bit back and forth between moving that error checking back into the `Fetch` vs returning an error from `NewPlainImage`

My main thought was we have 2 constructors, `NewPlainImage` and `NewFetchedPlainImageWithTag`. where `NewFetchedPlainImageWithTag` does panic if the provided ref is incorrect. Although *today*, due to 'when' and 'how' it is used, should *never* be in a position to be provided an incorrect ref (i think). Although we *could* introduce user validation around the `-i` flag and when reading image refs from the ImagesLock file to ensure that the ref passed into `NewPlainImage` is valid (just like its sibling NewFetchedPlainImageWithTag)

Went with removing the panic from `NewPlainImage` and moving the error back into `Fetch`.

However, I did add one more additional check in `NewFetchedPlainImageWithTag` to ensure that an image *is* in fact provided (and not nil), otherwise when `Fetch` is called, it won't return the fetched image, and instead the digest ref provided to the sibling constructor will be 'clobbered' by the 'unparsedRef' (which will be an empty string).

Reply:
got it. 👍